2021q3/copy_staging.adoc

```diff
@@ -0,0 +1,76 @@
+disk, CPU microcode update etc) at the contigous physical memory block
+starting at 2M.  This algorithm goes back to how i386 kernel boots.
+
+Also, when preparing to handle the control of to the kernel, loader
+creates quite special temporal mappings, where low 1G of physical
+address space is mapped 1:1 into virtual address space, and then
+repeated for each 1G until the virtual memory end.  Kernel knows about
+its physical location and the temporal mapping, and constructs kernel
+page tables assuming that the physical address of the text is at 2M.
+
+This mechanism of loader to kernel handoff was left unchanged when
+loader got support for UEFI environment.  Loader prepares kernel and
+auxillary preload data in so called staging area while UEFI boot
+services are active, and after EFI_BOOT_SERVICES.ExitBootServices(),
+the temporal mapping is activated and staging area is copied at 2M.
+
+An advantage at that time was that no changes to the kernel were
+needed.  But there are issues, the biggest is that memory at 2M might
+be not free for reuse.  For instance, BIOS runtime code or data might
+be located there.  Or there might be no memory at 2M at all.  Or
+trampoling page table or code, or even some parts of the staging area
+overlapped with the 2M region where staging area is copied.  The
+outcome was a weird and hard to diagnose boot time failures, typically
+machine hangs hard when loader started kernel.
+
+Another limitation is the 1G transient mapping, which due to copying
+means that total size of preloaded data cannot exceed around 400M for
+everything, including kernel, memory disks, and anything else.  Also
+the code to grow staging area on demand was quite unflexible, only
+able to grow the staging area in place.
+
+Described work allows UEFI loader on amd64 to start kernel from the
+staging area without copying.  Kernel assumptions about the hand-off
+were explicitly identified and documented.  Kernel only requires that
+the staging area was located below 4G together with the trampoline
+page table (this is a consequence of CPU architecture requiring 32bit
+protected mode to enter long mode), be 2M aligned and whole low 4G
+mapped 1:1 at hand-off.  Kernel deduces its physical address and
+builds kernel page tables accordingly.
+
+Making booting with staging area in-place work required identifying
+all places where amd64 kernel grown a dependency on its physical
+location.  Most complicated part was application processors startup,
+which required rewriting initialization code, which we were able to
+streamline as result.  In particular, when AP enters paging mode, it
+does so right into correct kernel page table, without loading
+intermediate trampoline page table.
+
+Updated loader automatically detects if the loaded kernel can handle
+in-place staging are ('non-copying mode').  If needed, this can be
+controlled by loader copy_staging command.  Also, the code to grow
```

Comment:
maybe "overridden with the ... command". Perhaps include an inline example if it doesn't make this too awkward.